token-swap/program/src/curve/stable.rs

```diff
@@ -0,0 +1,467 @@
+        let host_fee_denominator = 1;
+        let amp = 1;
+        let calculator = StableCurve {
+            trade_fee_numerator,
+            trade_fee_denominator,
+            owner_trade_fee_numerator,
+            owner_trade_fee_denominator,
+            owner_withdraw_fee_numerator,
+            owner_withdraw_fee_denominator,
+            host_fee_numerator,
+            host_fee_denominator,
+            amp,
+        };
+        assert_eq!(calculator.new_pool_supply(), INITIAL_SWAP_POOL_AMOUNT);
+    }
+
+    fn check_pool_token_rate(token_a: u128, deposit: u128, supply: u128, expected: Option<u128>) {
+        let trade_fee_numerator = 0;
+        let trade_fee_denominator = 1;
+        let owner_trade_fee_numerator = 0;
+        let owner_trade_fee_denominator = 1;
+        let owner_withdraw_fee_numerator = 0;
+        let owner_withdraw_fee_denominator = 1;
+        let host_fee_numerator = 0;
+        let host_fee_denominator = 1;
+        let amp = 1;
+        let calculator = StableCurve {
+            trade_fee_numerator,
+            trade_fee_denominator,
+            owner_trade_fee_numerator,
+            owner_trade_fee_denominator,
+            owner_withdraw_fee_numerator,
+            owner_withdraw_fee_denominator,
+            host_fee_numerator,
+            host_fee_denominator,
+            amp,
+        };
+        assert_eq!(
+            calculator.pool_tokens_to_trading_tokens(deposit, supply, token_a),
+            expected
+        );
+    }
+
+    #[test]
+    fn trading_token_conversion() {
+        check_pool_token_rate(2, 5, 10, Some(1));
+        check_pool_token_rate(10, 5, 10, Some(5));
+        check_pool_token_rate(5, 5, 10, Some(2));
+        check_pool_token_rate(5, 5, 10, Some(2));
+        check_pool_token_rate(u128::MAX, 5, 10, None);
+    }
+
+    #[test]
+    fn stable_swap_calculation_trade_fee() {
+        let swap_source_amount = 1_000;
+        let swap_destination_amount = 50_000;
+        let trade_fee_numerator = 1;
+        let trade_fee_denominator = 1_000;
+        let owner_trade_fee_numerator = 0;
+        let owner_trade_fee_denominator = 0;
+        let owner_withdraw_fee_numerator = 0;
+        let owner_withdraw_fee_denominator = 0;
+        let host_fee_numerator = 0;
+        let host_fee_denominator = 0;
+        let source_amount = 100;
+        let amp = 1;
+        let curve = StableCurve {
+            trade_fee_numerator,
+            trade_fee_denominator,
+            owner_trade_fee_numerator,
+            owner_trade_fee_denominator,
+            owner_withdraw_fee_numerator,
+            owner_withdraw_fee_denominator,
+            host_fee_numerator,
+            host_fee_denominator,
+            amp,
+        };
+        let result = curve
+            .swap(source_amount, swap_source_amount, swap_destination_amount)
+            .unwrap();
+        assert_eq!(result.new_source_amount, 1_100);
+        assert_eq!(result.amount_swapped, 2_063);
```

Comment:
Where do these numbers come from? Is there a reference calculation somewhere?

Reply:
It looks like these tests are just taken from `contant_product.rs` and the expected values are hardcoded to whatever values the `swap` function produces.

Math for this curve invariant is a lot more complicated compared to the constant product curve invariant. I still recommend porting over the thin simulation wrapper I wrote and add tests to check math against the simulation script provided by Curve.fi.

The range of numbers tested in these tests is really small and does not account for things like very big numbers, etc.

Reply:
Gotcha, yeah that would be great.  I wasn't sure if there were some published reference calculation numbers that we use to be sure we're doing the right thing.

Reply:
I propose to leave it as it is now and use the next PR to increase number of test cases and update to u256 math if needed (which is most likely)